### DISCUSSION TITLE
Existe algum algoritmo para validar se um NIF AO é válido. Um checkdigit?

### DISCUSSION THREAD

**Discussion by rmglsnts** [assoft-portugal/SAF-T-AO#132] [Q&A] [answered]
De forma a conseguirmos validar nos dados mestre de vários softwares, existe algum algoritmo de validação de nif angolano, que possibilite ser testado o nif introduzido pelo utilizador?

Obrigado.

**Comment by cryptolopes** [comment, marked as answer]
Ao contrário do que se sucede em diversas geografias, incluindo em Portugal, o `NIF` em Angola não pode ser validado através do método de `checkdigit`. É possível validar o `checkdigit` nalguns casos, em particular quando se tratam de empresas. Porém, existem casos em que o NIF é o mesmo que o número de identidade e outros ainda. Por essa razão não é possível aplicar o mesmo algoritmo em todos os casos.

### Checkdigit para empresas

📎 [NIF Angola (Validação).xlsx](https://github.com/assoft-portugal/SAF-T-AO/files/5998123/NIF.Angola.Validacao.xlsx)

### Métodos alternativos

#### Webservice

Possivelmente, um método alternativo é obter a informação sobre o NIF da própria AGT.
Em tempos falou-se de um `webservice` para o efeito, mas não sei se está aberto aos software builders ou sequer se ainda está em produção. Mais informação [aqui](https://t.me/saftao/2709) e [aqui](https://t.me/saftao/2710).

#### Webhook
Usar o serviço de consulta de NIF existente no site da AGT.

```text
url: https://agt.minfin.gov.ao/PortalAGT/#!/servicos/consultar-nif
```

Algumas dicas:
- [reCAPTCHA with node.js and express](https://stackoverflow.com/questions/49225216/recaptcha-with-node-js-and-express)
- [Using reCAPTCHA v3 with Node.js
](https://medium.com/@sergeisizov/using-recaptcha-v3-with-node-js-6a4b7bc67209)
- [Adding Recaptcha Token to Request in Nodejs](https://stackoverflow.com/questions/51429116/adding-recaptcha-token-to-request-in-nodejs)

**Reply by rmglsnts** [reply]
Obrigado pela pronta resposta.

**Reply by Kajibanga1993** [reply]
dizem que o número do billhete está actualmente vinculado ao número de NIf. assim sendo, caso seja verdade,você pode consultar  seu niff através do seu bi, implementando um serviço em sua Api que faça uma request do tipo get nesta url https://angolaapi.herokuapp.com/api/v1/validate/bi/[bi].
